Move version.txt to obj directory

chcosta opened the pull request:
Creating a version.txt file in the PackagesBasePath folder was not optimal because PackageBasePath contains the $(OSGroup) which gets filtered for a build.  The result is that you could have multiple PackageBasePaths, which meant generating redundant version.txt files.  We really just want to generate one version.txt file and use that for all of the produced packages.  In some repo's, like buildtools, this doesn't matter as much as in other repo's, like CoreFx, where there are multiple OSGroups for a build.  

This change moves the version.txt file to get generated in the obj directory, and creates a traversal build dependency so that the target gets called once before a traversal build.

@joperezr @weshaggard

weshaggard commented on src/Microsoft.DotNet.Build.Tasks/PackageFiles/versioning.targets:
```diff
@@ -217,13 +217,20 @@
+  </PropertyGroup>
+    
+  <PropertyGroup>
+    <SyncInfoDirectory>$(ObjDir)</SyncInfoDirectory>
```
Add conditions to allow these to be overridden if needed.

weshaggard commented:
other then the one override comment looks good to me.

chcosta commented:
@weshaggard , updated as per your comment.  I also changed the "TraversalBuildDependsOn" property to "BeforeTraversalBuildDependsOn" because, by convention, versioning.targets gets imported before dir.traversal.targets and I want to be able to call this target before "BuildAllProjects", https://github.com/dotnet/corefx/blob/master/dir.traversal.targets#L295.  The alternative would be to require versioning.targets to be imported after dir.traversal.targets and I didn't like that.

chcosta commented on src/Microsoft.DotNet.Build.Tasks/PackageFiles/versioning.targets:
```diff
@@ -217,13 +217,20 @@
 
   <PropertyGroup>
     <GetNuGetPackageVersionsDependsOn>$(GetNuGetPackageVersionsDependsOn);CreateVersionInfoFile</GetNuGetPackageVersionsDependsOn>
+    <BeforeTraversalBuildDependsOn>CreateVersionInfoFile;$(BeforeTraversalBuildDependsOn)</BeforeTraversalBuildDependsOn>
```
Your assessment is correct.  I like your suggestion.  I'll prepare an update

chcosta commented on src/Microsoft.DotNet.Build.Tasks/PackageFiles/versioning.targets:
```diff
@@ -218,12 +218,19 @@
+    
+  <PropertyGroup>
+    <SyncInfoDirectory Condition="'$(SyncInfoDirectory)' == ''">$(ObjDir)</SyncInfoDirectory>
+    <SyncInfoFile Condition="'$(SyncInfoFile)' == ''">$(SyncInfoDirectory)version.txt</SyncInfoFile>
```
I considered this, but for official builds, where we are shipping packages from, this will not be an issue.

joperezr replied:
agreed, but it will be an issue for local builds, although I guess people might not care if the binaries that they build might have an inaccurate version
